### ISSUE TITLE
2.8 - DISTINCT is removed from $query['fields'] if fields string contains newlines

### ISSUE THREAD

**Issue by steinkel** [cakephp/cakephp#8179]
While working on a custom finder I've used a field defined as
````php
$query['fields'] = array('Example.id', 'COUNT(
    DISTINCT (CASE
        WHEN Example.example_type = \'OPTION1\'
            THEN Example.unique_id
    END))');
````

I get the query:

````sql
SELECT `Model`.`id`, DISTINCT `COUNT( (CASE WHEN Model.example_type = 'OPTION1' THEN Model.unique_id END))` FROM ...
````

Note the DISTINCT is moved to the beginning of the query and the whole count expression passed as a column name, caused possibly by https://github.com/cakephp/cakephp/blob/2.8/lib/Cake/Model/Datasource/DboSource.php#L2568

If I reformat the query in 1 single line like this
````php
$query['fields'] = array('Example.id', 'COUNT(DISTINCT (CASE WHEN Example.example_type = \'OPTION1\' THEN Example.unique_id END))');
````

I get the following result:

````sql
SELECT `Example`.`id`, COUNT(DISTINCT (CASE WHEN Example.example_type = 'OPTION1' THEN Example.unique_id END))
````

Both uses cases should behave the same when building the query, or am I missing something?

I could provide a PR removing the newlines from the fields contents, if this makes sense for you.

**Comment by dereuromark** [comment]
Note that you should be using virtualFields for this, as a nice side effect this would also give you back a clean array with the right keys.
Could you check if the same happens when using virtualFields maybe?

**Comment by steinkel** [comment]
yep, I switched to virtualFields and works even with line breaks

**Comment by markstory** [comment]
Either a virtualField or expression object will work properly here.

> I could provide a PR removing the newlines from the fields contents, if this makes sense for you. 

That would be great.

**Comment by dereuromark** [comment]
Couldn't that potentially alter values that contain newlines (on purpose)?
I think there is nothing we can really do here without creating potential BC breaking other side-effects.
Especially since the intended behavior works.

**Comment by steinkel** [comment]
@dereuromark that's a good point, I would say replacing newlines would be a less probable use case but better we could provide a solution covering both cases...

**Comment by markstory** [comment]
@dereuromark Good point, I bet someone has included literal newlines in a CONCAT() somewhere. :trollface:

**Comment by steinkel** [comment]
Then I guess the best approach would be adding a note in the documentation?

**Comment by ravage84** [comment]
:+1: for the documentation approach.
